tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,339 @@
+    new_graph: Text, new_tensor_name: Text
+) -> Dict[Text, Dict[Text, tf.Tensor]]:
+  element = copy.deepcopy(element)
+  if new_graph not in element:
+    element[new_graph] = {}
+  element[new_graph][new_tensor_name] = element[old_graph][old_tensor_name]
+  return element
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(Dict[Text, Dict[Text, tf.Tensor]])
+@beam.typehints.with_output_types(Dict[Text, Dict[Text, tf.Tensor]])
+def _ExtractRemoteGraphOutput(  # pylint: disable=invalid-name
+    pcoll: beam.pvalue.PCollection,
+    parent_remote_op_name: Text,
+    child_remote_op_name: Text,
+    remote_op_name_to_graph_name: Mapping[Text, Text],
+    graph_name_to_specs: Mapping[Text, List[execution_spec.ExecutionSpec]],
+) -> beam.pvalue.PCollection:
+  """Extracts remote graph output.
+
+  After finish executing a remote graph, we need to collect its output.
+  We first find the output name of the remote graph, then we copy the
+  output of the remote graph to its parent graph. Finally, we clear the
+  intermediate results of the remote graph.
+
+  Note we assumed that each node has only one output, which also applies
+  to remote op. This means that a remote graph can only have one output.
+
+  Args:
+    pcoll: A PCollection of child graph results. Each element is a dictionary
+           from remote op names to a dictionary from tensor names to tensors.
+           Here, element[child_remote_op_name] stores graph inputs,
+           intermediate results, and graph output.
+    parent_remote_op_name: The remote op name of the parent graph.
+    child_remote_op_name: The remote op name of the child graph.
+    remote_op_name_to_graph_name:
+      A mapping from remote op names to graph names.
+    graph_name_to_specs: A mapping from graph names to a list of
+                         ExecutionSpecs.
+
+  Returns:
+    A PCollection of child graph output in parent graph. Each element is a
+    dictionary from remote op names to a dictionary from tensor names to
+    tensors. Here, element[parent_remote_op_name] contains the output from
+    the child graph, and element[child_remote_op_name] becomes empty.
+  """
+  child_graph_name = remote_op_name_to_graph_name[child_remote_op_name]
+  child_specs = graph_name_to_specs[child_graph_name]
+  child_output_name = list(child_specs[-1].output_names)[0]
+
+  step_name_extract = ("ExtractOutput[Graph_%s][Output_%s]" %
+                       (child_remote_op_name, child_output_name))
+  step_name_clear = ("ClearIntermediateOutputs[Graph_%s]" %
+                     (child_remote_op_name))
+
+  return (pcoll
+          | step_name_extract >> beam.Map(
+              _copy_tensor,
+              child_remote_op_name,
+              _import_tensor_name(child_output_name),
+              parent_remote_op_name,
+              _import_tensor_name(child_remote_op_name))
+          | step_name_clear >> beam.Map(
+              _clear_outputs_for_finished_graph,
+              child_remote_op_name))
```

Comment:
What happens if we do not clear outputs? Will the re-execution of the same remote graph in later stages be affected?

Reply:
If we do not clear the outputs, the re-execution of the same remote graph will overwrite everything in element[child_remote_op_name].

Reply:
I agree it's clean to clear (with the current approach). But suppose we did not clear, and the re-execution will overwrite, but is correctness affected?

The general concern I have is that this kind of mutation may make it harder to parallelize the stages in the future. I wonder if we could have avoided it if we created a separate PCollection with elements that are discarded at the end (without affecting the main PCollection).

Reply:
The re-execution will overwrite the graph inputs and intermediate results for a remote graph, so the correctness wouldn't be affected.

Yes, I agree that this kind of mutation makes it harder to parallel between remote ops inside the same remote op layer. This morning, I tried to replace element from {remote op name: {tensor name: tensor}} to {tensor name: tensor}, where each element only caches the intermediate results of one graph. Indeed, we can simplify some abstractions by doing this. However, I encountered a bug. The bug is that {tensor name: tensor} becomes {tensor name: tensor name} after some stage, but I couldn't find the cause of it.

Reply:
Not necessary to change.